File: objc/objcfind/Sources/objcfind/FindSettings.m
#import "common.h"
#import "FileTypes.h"
#import "FindSettings.h"
#import <objc/runtime.h>

@implementation FindSettings

@synthesize archivesOnly = _archivesOnly;
@synthesize debug = _debug;

- (instancetype) init {
    self = [super init];
    if (self) {
        self.archivesOnly = false;
        self.colorize = true;
        self.debug = false;
        self.defaultFiles = true;
        self.dirColor = ColorCyan;
        self.extColor = ColorYellow;
        self.fileColor = ColorMagenta;
        self.followSymlinks = false;
        self.includeArchives = false;
        self.includeHidden = false;
        self.printDirs = false;
        self.printFiles = false;
        self.printUsage = false;
        self.printVersion = false;
        self.recursive = true;
        self.sortCaseInsensitive = false;
        self.sortDescending = false;
        self.verbose = false;

        self.maxDepth = -1;
        self.maxLastMod = nil;
        self.maxSize = 0;
        self.minDepth = -1;
        self.minLastMod = nil;
        self.minSize = 0;

        self.inArchiveExtensions = [[NSMutableArray alloc] init];
        self.inArchiveFilePatterns = [[NSMutableArray alloc] init];
        self.inDirPatterns = [[NSMutableArray alloc] init];
        self.inExtensions = [[NSMutableArray alloc] init];
        self.inFilePatterns = [[NSMutableArray alloc] init];
        self.inFileTypes = [[NSMutableArray alloc] init];
        self.outArchiveExtensions = [[NSMutableArray alloc] init];
        self.outArchiveFilePatterns = [[NSMutableArray alloc] init];
        self.outDirPatterns = [[NSMutableArray alloc] init];
        self.outExtensions = [[NSMutableArray alloc] init];
        self.outFilePatterns = [[NSMutableArray alloc] init];
        self.outFileTypes = [[NSMutableArray alloc] init];
        self.paths = [[NSMutableArray alloc] init];
        self.sortBy = SortByFilePath;
    }
    return self;
}

- (NSString *) description {

    unsigned int count;
    // Get the list of all properties for the class
    objc_property_t *properties = class_copyPropertyList([self class], &count);

    NSMutableDictionary *propDict = [[NSMutableDictionary alloc] initWithCapacity:count];

    for (unsigned int i = 0; i < count; i++) {
        objc_property_t property = properties[i];
        // Get the property name (label)
        const char *propertyName = property_getName(property);
        NSString *name = [NSString stringWithUTF8String:propertyName];

        // Get the property value using Key-Value Coding (KVC)
        id value = [self valueForKey:name];

        if (value == nil) {
            value = @"0";
        }
        propDict[name] = value;
    }
    
    // Free the list of properties
    free(properties);

    NSMutableString *d = [[NSMutableString alloc] initWithString:@"ZZFindSettings("];

    NSArray<NSString*> *keys = [[propDict allKeys] sortedArrayUsingSelector:@selector(compare:)];
    int idx = 0;
    for (NSString *key in keys) {
        NSObject *val = propDict[key];
        if (idx > 0) {
            [d appendString:@", "];
        }
        if ([key hasSuffix:@"Color"]) {
            NSNumber *num = (NSNumber *)val;
            NSInteger i = [num integerValue];
            NSString *colorName = getNameFromColor((int)i);
            [d appendFormat:@"%@=%@", key, colorName];
        } else if ([key hasSuffix:@"Depth"] || [key hasSuffix:@"Size"]) {
            NSNumber *num = (NSNumber *)val;
            [d appendFormat:@"%@=%@", key, num];
        } else if ([key hasSuffix:@"LastMod"]) {
            if ([val isKindOfClass:[NSString class]]) {
                [d appendFormat:@"%@=%@", key, val];
            } else {
                NSDate *date = (NSDate *)val;
                [d appendFormat:@"%@=%@", key, [FindSettings lastModToNSString:date]];
            }
        } else if ([key isEqualToString:@"sortBy"]) {
            NSNumber *num = (NSNumber *)val;
            NSInteger i = [num integerValue];
            NSString *sortByName = [FindSettings getNameFromSortBy:(int)i];
            [d appendFormat:@"%@=%@", key, sortByName];

        } else if ([val isKindOfClass:[NSArray class]]) {
            NSArray *arr = (NSArray *)val;
            if ([key hasSuffix:@"FileTypes"]) {
                [d appendFormat:@"%@=%@", key, [FindSettings fileTypesArrayToNSString:arr]];
            } else {
                [d appendFormat:@"%@=%@", key, arrayToNSString(arr)];
            }
        } else {
            NSNumber *num = (NSNumber *)val;
            BOOL b = [num boolValue];
            [d appendFormat:@"%@=%@", key, boolToNSString(b)];
        }

        idx++;
    }

    [d appendString:@")"];
    return d;
}

- (void) addExtensions:(NSString *)ext toArr:(NSMutableArray *)arr {
    NSArray *exts = [ext componentsSeparatedByString:@","];
    [arr addObjectsFromArray:exts];
}

- (void) addPattern:(NSString *)pattern toArr:(NSMutableArray *)arr {
    [arr addObject:[[Regex alloc] initWithPattern:pattern]];
}

- (void) addInArchiveExtension:(NSString *)ext {
    [self addExtensions:ext toArr:self.inArchiveExtensions];
}

- (void) addInArchiveFilePattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.inArchiveFilePatterns];
}

- (void) addInDirPattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.inDirPatterns];
}

- (void) addInExtension:(NSString *)ext {
    [self addExtensions:ext toArr:self.inExtensions];
}

- (void) addInFilePattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.inFilePatterns];
}

- (void) addOutArchiveExtension:(NSString *)ext {
    [self addExtensions:ext toArr:self.outArchiveExtensions];
}

- (void) addOutArchiveFilePattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.outArchiveFilePatterns];
}

- (void) addOutDirPattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.outDirPatterns];
}

- (void) addOutExtension:(NSString *)ext {
    [self addExtensions:ext toArr:self.outExtensions];
}

- (void) addOutFilePattern:(NSString *)pattern {
    [self addPattern:pattern toArr:self.outFilePatterns];
}

- (void) addFileType:(NSString *)typeName toArr:(NSMutableArray *)arr {
    FileType fileType = [FileTypes fromName:typeName];
    [arr addObject:[NSNumber numberWithInt:fileType]];
    // TODO: if we do this for objc/swift we need to do it for all of other languages
    // if fileType is FileTypeText, add text sub-types
//    if (fileType == FileTypeText) {
//        [arr addObject:[NSNumber numberWithInt:FileTypeCode]];
//        [arr addObject:[NSNumber numberWithInt:FileTypeXml]];
//    }
}

- (void) addInFileType:(NSString *)typeName {
    [self addFileType:typeName toArr:self.inFileTypes];
}

- (void) addOutFileType:(NSString *)typeName {
    [self addFileType:typeName toArr:self.outFileTypes];
}

- (void) addPath:(NSString *)path {
    [self.paths addObject:path];
}

- (BOOL) archivesOnly {
    return _archivesOnly;
}

- (void)setArchivesOnly:(BOOL)b {
    _archivesOnly = b;
    if (b) {
        [self setIncludeArchives:b];
    }
}

- (BOOL) debug {
    return _debug;
}

- (void)setDebug:(BOOL)b {
    _debug = b;
    if (b) {
        [self setVerbose:b];
    }
}

- (void) setMaxDepthFromString:(NSString*)depthStr {
    [self setMaxDepth:[depthStr intValue]];
}

- (void) setMaxLastModFromString:(NSString*)dateStr {
    [self setMaxLastMod:stringToNSDate(dateStr)];
}

- (void) setMaxSizeFromString:(NSString*)sizeStr {
    [self setMaxSize:[sizeStr intValue]];
}

- (void) setMinDepthFromString:(NSString*)depthStr {
    [self setMinDepth:[depthStr intValue]];
}

- (void) setMinLastModFromString:(NSString*)dateStr {
    [self setMinLastMod:stringToNSDate(dateStr)];
}

- (void) setMinSizeFromString:(NSString*)sizeStr {
    [self setMinSize:[sizeStr intValue]];
}

+ (SortBy)getSortByFromName:(NSString *)sortByName {
    NSString *lname = [sortByName lowercaseString];
    if (lname == [NSString stringWithUTF8String:S_FILENAME] || lname == [NSString stringWithUTF8String:S_NAME]) {
        return SortByFileName;
    }
    if (lname == [NSString stringWithUTF8String:S_FILESIZE] || lname == [NSString stringWithUTF8String:S_SIZE]) {
        return SortByFileSize;
    }
    if (lname == [NSString stringWithUTF8String:S_FILETYPE] || lname == [NSString stringWithUTF8String:S_TYPE]) {
        return SortByFileType;
    }
    if (lname == [NSString stringWithUTF8String:S_LASTMOD]) {
        return SortByLastMod;
    }
    return SortByFilePath;
}

+ (NSString*) getNameFromSortBy:(SortBy)sortBy {
    if (sortBy == SortByFileName) {
        return [NSString stringWithUTF8String:S_FILENAME];
    }
    if (sortBy == SortByFileSize) {
        return [NSString stringWithUTF8String:S_FILESIZE];
    }
    if (sortBy == SortByFileType) {
        return [NSString stringWithUTF8String:S_FILETYPE];
    }
    if (sortBy == SortByLastMod) {
        return [NSString stringWithUTF8String:S_LASTMOD];
    }
    return [NSString stringWithUTF8String:S_FILEPATH];
}

- (void) setSortByFromName:(NSString*)sortByName {
    self.sortBy = [FindSettings getSortByFromName:sortByName];
}

+ (NSString*) lastModToNSString:(NSDate *)lastMod {
    if (lastMod == nil) {
        return @"0";
    }
    return [NSString stringWithFormat:@"\"%@\"", dateToNSString(lastMod)];
}

+ (NSString*) fileTypesArrayToNSString:(NSArray<NSNumber*>*)arr {
    NSMutableString *arrString = [NSMutableString stringWithString:@"["];
    for (int i=0; i < [arr count]; i++) {
        if (i > 0) {
            [arrString appendString:@", "];
        }
        NSString *typeName = [FileTypes toName:[arr[i] intValue]];
        [arrString appendFormat:@"%@", typeName];
    }
    [arrString appendString:@"]"];
    return [NSString stringWithString:arrString];
}

- (BOOL) needLastMod {
    return self.sortBy == SortByLastMod ||
    self.maxLastMod != nil || self.minLastMod != nil;
}

- (BOOL) needSize {
    return self.sortBy == SortByFileSize ||
    self.maxSize > 0 || self.minSize > 0;
}

@end

File: swift/swiftfind/Sources/swiftfind/FindSettings.swift
//
//  FindSettings.swift
//  swiftfind
//
//  Created by Cary Clark on 5/12/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation

private let sortByFilePathName = "filepath"
private let sortByFileNameName = "filename"
private let sortByNameName = "name"
private let sortByFileSizeName = "filesize"
private let sortBySizeName = "size"
private let sortByFileTypeName = "filetype"
private let sortByTypeName = "type"
private let sortByLastModName = "lastmod"

public enum SortBy {
    case filePath, fileName, fileSize, fileType, lastMod
}

public func nameToSortBy(_ sortByName: String) -> SortBy {
    let lname = sortByName.lowercased()
    switch lname {
    case sortByFileNameName, sortByNameName:
        return SortBy.fileName
    case sortByFileSizeName, sortBySizeName:
        return SortBy.fileSize
    case sortByFileTypeName, sortByTypeName:
        return SortBy.fileType
    case sortByLastModName:
        return SortBy.lastMod
    default:
        return SortBy.filePath
    }
}

public func sortByToName(_ sortBy: SortBy) -> String {
    switch sortBy {
    case SortBy.fileName:
        sortByFileNameName
    case SortBy.fileSize:
        sortByFileSizeName
    case SortBy.fileType:
        sortByFileTypeName
    case SortBy.lastMod:
        sortByLastModName
    default:
        sortByFilePathName
    }
}

public enum DefaultFindSettings {
    public static let archivesOnly = false
    public static let colorize = true
    public static let debug = false
    public static let dirColor = Color.cyan
    public static let extColor = Color.yellow
    public static let fileColor = Color.magenta
    public static let followSymlinks = false
    public static let includeArchives = false
    public static let includeHidden = false
    public static let maxDepth: Int32 = -1
    public static let maxLastMod: Date? = nil
    public static let maxSize: UInt64 = 0
    public static let minDepth: Int32 = -1
    public static let minLastMod: Date? = nil
    public static let minSize: UInt64 = 0
    public static let printDirs = false
    public static let printFiles = false
    public static let printUsage = false
    public static let printVersion = false
    public static let recursive = true
    public static let sortCaseInsensitive = false
    public static let sortDescending = false
    public static let verbose = false
}

open class FindSettings: CustomStringConvertible {
    // these are public because they must be accessible outside the package
    public var _archivesOnly: Bool = DefaultFindSettings.archivesOnly
    open var colorize: Bool = DefaultFindSettings.colorize
    public var _debug: Bool = DefaultFindSettings.debug
    open var dirColor: Color = DefaultFindSettings.dirColor
    open var extColor: Color = DefaultFindSettings.extColor
    open var fileColor: Color = DefaultFindSettings.fileColor
    open var followSymlinks: Bool = DefaultFindSettings.followSymlinks
    open var includeArchives: Bool = DefaultFindSettings.includeArchives
    open var includeHidden: Bool = DefaultFindSettings.includeHidden
    open var maxDepth: Int32 = DefaultFindSettings.maxDepth
    open var maxLastMod: Date? = DefaultFindSettings.maxLastMod
    open var maxSize: UInt64 = DefaultFindSettings.maxSize
    open var minDepth: Int32 = DefaultFindSettings.minDepth
    open var minLastMod: Date? = DefaultFindSettings.minLastMod
    open var minSize: UInt64 = DefaultFindSettings.minSize
    open var printDirs: Bool = DefaultFindSettings.printDirs
    open var printFiles: Bool = DefaultFindSettings.printFiles
    open var printUsage: Bool = DefaultFindSettings.printUsage
    open var printVersion: Bool = DefaultFindSettings.printVersion
    open var recursive: Bool = DefaultFindSettings.recursive
    open var sortCaseInsensitive: Bool = DefaultFindSettings.sortCaseInsensitive
    open var sortDescending: Bool = DefaultFindSettings.sortDescending
    open var verbose: Bool = DefaultFindSettings.verbose

    open var inArchiveExtensions = Set<String>()
    open var inArchiveFilePatterns = [Regex]()
    open var inDirPatterns = [Regex]()
    open var inExtensions = Set<String>()
    open var inFilePatterns = [Regex]()
    open var inFileTypes = [FileType]()
    open var outArchiveExtensions = Set<String>()
    open var outArchiveFilePatterns = [Regex]()
    open var outDirPatterns = [Regex]()
    open var outExtensions = Set<String>()
    open var outFilePatterns = [Regex]()
    open var outFileTypes = [FileType]()
    open var paths = Set<String>()
    open var sortBy = SortBy.filePath

    public init() {}

    func splitExtensions(_ exts: String) -> [String] {
        exts.split { $0 == "," }.map { String($0) }
    }

    public func addInArchiveExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            inArchiveExtensions.insert(ext)
        }
    }

    public func addInExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            inExtensions.insert(ext)
        }
    }

    public func addInArchiveFilePattern(_ pattern: String) {
        inArchiveFilePatterns.append(Regex(pattern))
    }

    public func addInDirPattern(_ pattern: String) {
        inDirPatterns.append(Regex(pattern))
    }

    public func addInFilePattern(_ pattern: String) {
        inFilePatterns.append(Regex(pattern))
    }

    public func addInFileType(_ typeName: String) {
        let fileType = FileTypes.fromName(typeName)
        inFileTypes.append(fileType)
        // TODO: if we do this for objc/swift we need to do it for all of other languages
        // if text, add text sub-types
//        if fileType == FileType.text {
//            inFileTypes.append(FileType.code)
//            inFileTypes.append(FileType.xml)
//        }
    }

    public func addOutArchiveExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            outArchiveExtensions.insert(ext)
        }
    }

    public func addOutExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            outExtensions.insert(ext)
        }
    }

    public func addOutArchiveFilePattern(_ pattern: String) {
        outArchiveFilePatterns.append(Regex(pattern))
    }

    public func addOutDirPattern(_ pattern: String) {
        outDirPatterns.append(Regex(pattern))
    }

    public func addOutFilePattern(_ pattern: String) {
        outFilePatterns.append(Regex(pattern))
    }

    public func addOutFileType(_ typeName: String) {
        let fileType = FileTypes.fromName(typeName)
        outFileTypes.append(fileType)
        // TODO: if we do this for objc/swift we need to do it for all of other languages
        // if text, add text sub-types
//        if fileType == FileType.text {
//            outFileTypes.append(FileType.code)
//            outFileTypes.append(FileType.xml)
//        }
    }

    public func setMaxDepthFromString(_ maxDepthStr: String) {
        maxDepth = Int32(maxDepthStr) ?? 0
    }

    public func setMaxLastModFromString(_ maxLastModStr: String) {
        maxLastMod = stringToDate(maxLastModStr)
    }

    public func setMaxSizeFromString(_ maxSizeStr: String) {
        maxSize = UInt64(maxSizeStr) ?? 0
    }

    public func setMinDepthFromString(_ minDepthStr: String) {
        minDepth = Int32(minDepthStr) ?? 0
    }

    public func setMinLastModFromString(_ minLastModStr: String) {
        minLastMod = stringToDate(minLastModStr)
    }

    public func setMinSizeFromString(_ minSizeStr: String) {
        minSize = UInt64(minSizeStr) ?? 0
    }

    public func needLastMod() -> Bool {
        sortBy == SortBy.lastMod ||
            maxLastMod != nil || minLastMod != nil
    }

    public func needSize() -> Bool {
        sortBy == SortBy.fileSize ||
            maxSize > 0 || minSize > 0
    }

    public func addPath(_ path: String) {
        paths.insert(path)
    }

    public func setSortBy(_ sortByName: String) {
        sortBy = nameToSortBy(sortByName)
    }

    open var archivesOnly: Bool {
        get {
            _archivesOnly
        }
        set {
            _archivesOnly = newValue
            if newValue {
                includeArchives = newValue
            }
        }
    }

    open var debug: Bool {
        get {
            _debug
        }
        set {
            _debug = newValue
            if newValue {
                verbose = newValue
            }
        }
    }

    open var description: String {
        "FindSettings(" +
            "archivesOnly=\(archivesOnly)" +
            ", colorize=\(colorize)" +
            ", debug=\(debug)" +
            ", followSymlinks=\(followSymlinks)" +
            ", inArchiveExtensions=\(setToString(inArchiveExtensions))" +
            ", inArchiveFilePatterns=\(arrayToString(inArchiveFilePatterns))" +
            ", inDirPatterns=\(arrayToString(inDirPatterns))" +
            ", inExtensions=\(setToString(inExtensions))" +
            ", inFilePatterns=\(arrayToString(inFilePatterns))" +
            ", inFileTypes=\(arrayToString(inFileTypes.map { FileTypes.toName($0) }, false))" +
            ", includeArchives=\(includeArchives)" +
            ", includeHidden=\(includeHidden)" +
            ", maxDepth=\(maxDepth)" +
            ", maxLastMod=\(dateToString(maxLastMod))" +
            ", maxSize=\(maxSize)" +
            ", minDepth=\(minDepth)" +
            ", minLastMod=\(dateToString(minLastMod))" +
            ", minSize=\(minSize)" +
            ", outArchiveExtensions=\(setToString(outArchiveExtensions))" +
            ", outArchiveFilePatterns=\(arrayToString(outArchiveFilePatterns))" +
            ", outDirPatterns=\(arrayToString(outDirPatterns))" +
            ", outExtensions=\(setToString(outExtensions))" +
            ", outFilePatterns=\(arrayToString(outFilePatterns))" +
            ", outFileTypes=\(arrayToString(outFileTypes.map { FileTypes.toName($0) }, false))" +
            ", paths=\(setToString(paths))" +
            ", printDirs=\(printDirs)" +
            ", printFiles=\(printFiles)" +
            ", printUsage=\(printUsage)" +
            ", printVersion=\(printVersion)" +
            ", recursive=\(recursive)" +
            ", sortBy=\(sortByToName(sortBy))" +
            ", sortCaseInsensitive=\(sortCaseInsensitive)" +
            ", sortDescending=\(sortDescending)" +
            ", verbose=\(verbose)" +
            ")"
    }
}
